File: app/src/main/kotlin/com/spacex/ui/rocketList/RocketListContract.java
package com.spacex.ui.rocketList;

import androidx.annotation.NonNull;
import java.util.List;

interface RocketListContract {

  interface View {

    void setUpViews();

    void showAsLoading();

    void showAsEmpty();

    void showAsErrorLoading();

    void showRockets(@NonNull final List<RocketUIM> rockets);

    void showAllRocketsFilterAsSelected();

    void showActiveRocketsFilterAsSelected();

    void navigateToRocketDetails(final int rocketId);

    void navigateToWelcomeActivity();

    void navigateToAboutActivity();

    void showFilterOptions();

    void hideFilterOptions();

    void navigateBack();
  }

  interface Presenter {

    void onCreate();

    void onStart();

    void onRocketClicked(final int rocketId);

    void onFilterRocketsClick();

    void onShowAllRocketsFilterOptionClick();

    void onShowActiveRocketsFilterOptionClick();

    void onCloseFilterOptions();

    void onShowWelcomeClick();

    void onAboutWelcomeClick();

    void onRefresh();

    void onRetry();

    void onBackNavigationPressed();

    void onStop();

    void onDestroy();
  }

  interface Coordinator {
    void loadAllRockets(@NonNull final OnLoadRocketListCallbacks callbacks);

    void loadActiveRockets(@NonNull final OnLoadRocketListCallbacks callbacks);

    void cancel();

    interface OnLoadRocketListCallbacks {
      void onSuccessLoadingRocketList(@NonNull final List<RocketUIM> rockets);

      void onErrorLoadingRocketList();
    }
  }
}

File: app/src/main/kotlin/com/spacex/ui/rocketList/RocketListContract.kt
package com.spacex.ui.rocketList

interface RocketListContract {

    interface View {

        fun setUpViews()

        fun showAsLoading()

        fun showAsEmpty()

        fun showAsErrorLoading()

        fun showRockets(rockets: List<RocketUIM>)

        fun showAllRocketsFilterAsSelected()

        fun showActiveRocketsFilterAsSelected()

        fun navigateToRocketDetails(rocketId: Int)

        fun navigateToWelcomeActivity()

        fun navigateToAboutActivity()

        fun showFilterOptions()

        fun hideFilterOptions()

        fun navigateBack()
    }

    interface Presenter {

        fun onCreate()

        fun onStart()

        fun onRocketClicked(rocketId: Int)

        fun onFilterRocketsClick()

        fun onShowAllRocketsFilterOptionClick()

        fun onShowActiveRocketsFilterOptionClick()

        fun onCloseFilterOptions()

        fun onShowWelcomeClick()

        fun onAboutWelcomeClick()

        fun onRefresh()

        fun onRetry()

        fun onBackNavigationPressed()

        fun onStop()

        fun onDestroy()
    }

    interface Coordinator {
        fun loadAllRockets(callbacks: OnLoadRocketListCallbacks)

        fun loadActiveRockets(callbacks: OnLoadRocketListCallbacks)

        fun cancel()

        interface OnLoadRocketListCallbacks {
            fun onSuccessLoadingRocketList(rockets: List<RocketUIM>)

            fun onErrorLoadingRocketList()
        }
    }
}
